### ISSUE TITLE
Array shape is lost during as_ctypes/as_array round trip in 1.15

### ISSUE THREAD

**Issue by Strilanc** [numpy/numpy#11636]
It appears that, when a 2 dimensional array's first index has a length of 1, round-tripping the array through `np.ctypeslib.as_ctypes` and `np.ctypeslib.as_array` loses the shape of the array.

Here is a short repro:

```python
import numpy as np
arr = np.array([[1, 2]])
c_arr = np.ctypeslib.as_ctypes(arr)
back = np.ctypeslib.as_array(c_arr)
np.testing.assert_equal(arr, back)

# In numpy 1.15:
# E       AssertionError: 
# E       Arrays are not equal
# E       
# E       (shapes (1, 2), (2,) mismatch)
# E        x: array([[1, 2]])
# E        y: array([1, 2])
```

Note that this did not occur in numpy v1.14. Another example of an array that doesn't round trip correctly is `[[1], [2]]`. However, `[[1, 2], [3, 4]]` does round trip correctly.

Note that a possibly-relevant warning which I do not understand occurs when executing the repro code:

```
RuntimeWarning: Item size computed from the PEP 3118 buffer format string does not match the actual item size.
```

**Comment by eric-wieser** [comment]
>  a possibly-relevant warning which I do not understand occurs

This warning is telling you that your version of python has a broken `ctypes` module. Unfortunately, there is no version of python yet released with a PEP3118-compliant ctypes module...

**Comment by eric-wieser** [comment]
I can't reproduce this - what is `sys.version_info` on your platform?

Can you show the result of `print(repr(c_arr))` and of `memoryview(c_arr).format`?

**Comment by Strilanc** [comment]
```
In [4]: print(
   ...:   memoryview(c_arr).itemsize,
   ...:   memoryview(c_arr).shape,
   ...:   np.array(c_arr).dtype,
   ...:   np.array(c_arr).shape,
   ...:   np.asarray(c_arr).dtype,
   ...:   np.asarray(c_arr).shape, sep='\n')
8
(1, 2)
int64
(1, 2)
int64
(1, 2)
```

Also some warnings were printed:

```
bin/ipython:4: RuntimeWarning: Item size computed from the PEP 3118 buffer format string does not match the actual item size.
  import re
bin/ipython:5: RuntimeWarning: Item size computed from the PEP 3118 buffer format string does not match the actual item size.
  import sys
numpy/core/numeric.py:492: RuntimeWarning: Item size computed from the PEP 3118 buffer format string does not match the actual item size.
  return array(a, dtype, copy=False, order=order)
```

**Comment by Strilanc** [comment]
I initially put 1.14 output by accident; you may have received a confusing email.

**Comment by eric-wieser** [comment]
I don't understand the warnings are being emitted there, but I do understand why the guess is failing:
```
>>> c_type_square = ctypes.c_uint8 * 3 * 3
>>> np.dtype(c_type_square)
dtype((('u1', (3,)), (3,)))  # a little janky, but array of arrays is a fair description
>>> c_type_col = ctypes.c_uint8 * 3 * 1
>>> np.dtype(c_type_col)
dtype(('u1', (3,))) # uh oh, singleton dimension is gone
>>> c_type_row = ctypes.c_uint8 * 1 * 3
>>> np.dtype(c_type_row)
dtype(('u1', (3,))) # same
```
This is caused by `np.dtype((scalar, N))` not working when `N == 1`. There is an issue about that somewhere, but I think we might be stuck with it.

**Comment by eric-wieser** [comment]
Patched at #11652 - thanks for the bug report!
